packages/next/src/server/web/adapter.ts

```diff
@@ -177,13 +184,42 @@ export async function adapter(
   }
 
   const event = new NextFetchEvent({ request, page: params.page })
-  let response = await params.handler(request, event)
+  let response
+  let cookiesFromResponse
+
+  // we only care to make async storage available for middleware
+  if (params.page === '/middleware') {
+    response = await RequestAsyncStorageWrapper.wrap(
+      requestAsyncStorage,
+      {
+        req: request,
+        renderOpts: {
+          onUpdateCookies: (cookies) => {
+            cookiesFromResponse = cookies
+          },
+          // @ts-expect-error: TODO: investigate why previewProps isn't on RenderOpts
+          previewProps: prerenderManifest?.preview || {
```

Comment:
Looks like we didn't add it as optional type here https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render/types.ts#L140

But this was already the case so can be updated separate.